packages/shopify-app-remix/src/server/authenticate/admin/authenticate.ts

```diff
@@ -68,16 +68,18 @@ export function authStrategyFactory<
   function createContext(
     request: Request,
     session: Session,
+    authStrategy: AuthorizationStrategy,
     sessionToken?: JwtPayload,
   ): AdminContext<ConfigArg, Resources> {
     const context:
       | EmbeddedAdminContext<ConfigArg, Resources>
       | NonEmbeddedAdminContext<ConfigArg, Resources> = {
-      admin: createAdminApiContext<Resources>(request, session, {
-        api,
-        logger,
-        config,
-      }),
+      admin: createAdminApiContext<Resources>(
+        request,
+        session,
+        params,
+        authStrategy,
```

Comment:
I like the solution of having the strategy determine how we handle errors, but I'm a little on the fence about passing in the entire auth strategy as a parameter for creating the context, it feels like we're mixing concerns a little bit.

What if, instead of passing the strategy in to `handleClientErrorFactory`, we have the strategy export its own flavour of `handleClientErrorFactory`, something like

```ts
  public async handleClientError({
    request,
    session,
  }: StrategyHandleClientErrorOptions): Promise<HandleAdminClientError> {
    const {config, api, logger} = this;
    return handleClientErrorFactory({
      onError: // Do token exchange things
    });
  }
```

That way, each strategy can determine how it handles the errors and they all subscribe to a common interface without the strategy having to be shared around.

IIRC `handleClientErrorFactory` is pretty much tied to the strategy already anyway, so it shouldn't really affect anything else. Perhaps it makes more sense for the strategy to just export a callback for its own error handling.

Does that sound sensible?